grandpa-client/verifier/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+
+mod state_machine;
+
+extern crate alloc;
+
+use alloc::vec;
+use anyhow::anyhow;
+use codec::{Decode, Encode};
+use finality_grandpa::Chain;
+use hash_db::Hasher;
+use primitives::{
+    error,
+    justification::{find_scheduled_change, AncestryChain, GrandpaJustification},
+    parachain_header_storage_key, ClientState, HostFunctions, ParachainHeaderProofs,
+    ParachainHeadersWithFinalityProof,
+};
+use sp_core::H256;
+use sp_runtime::traits::Header;
+use sp_trie::{LayoutV0, StorageProof};
+
+/// This function verifies the GRANDPA finality proof for relay chain headers.
+///
+/// Next, we prove the finality of parachain headers, by verifying patricia-merkle trie state proofs
+/// of these headers, stored at the recently finalized relay chain heights.
+pub fn verify_parachain_headers_with_grandpa_finality_proof<H, Host>(
+    mut client_state: ClientState,
+    proof: ParachainHeadersWithFinalityProof<H>,
+) -> Result<ClientState, error::Error>
+    where
+        H: Header<Hash = H256, Number = u32>,
+        H::Number: finality_grandpa::BlockNumberOps + Into<u32>,
+        Host: HostFunctions,
+        Host::BlakeTwo256: Hasher<Out = H256>,
+{
+    let ParachainHeadersWithFinalityProof { finality_proof, parachain_headers } = proof;
+
+    // 1. First validate unknown headers.
+    let headers = AncestryChain::<H>::new(&finality_proof.unknown_headers);
+
+    let target = finality_proof
+        .unknown_headers
+        .iter()
+        .max_by_key(|h| *h.number())
+        .ok_or_else(|| anyhow!("Unknown headers can't be empty!"))?;
+
+    // this is illegal
+    if target.hash() != finality_proof.block {
+        Err(anyhow!("Latest finalized block should be highest block in unknown_headers"))?;
+    }
+
+    let justification = GrandpaJustification::<H>::decode(&mut &finality_proof.justification[..])?;
+
+    if justification.commit.target_hash != finality_proof.block {
+        Err(anyhow!("Justification target hash and finality proof block hash mismatch"))?;
+    }
+
+    let from = client_state.latest_relay_hash;
+
+    let base = finality_proof
+        .unknown_headers
+        .iter()
+        .min_by_key(|h| *h.number())
+        .ok_or_else(|| anyhow!("Unknown headers can't be empty!"))?;
+
+    if base.number() < &client_state.latest_relay_height {
+        headers.ancestry(base.hash(), client_state.latest_relay_hash).map_err(|_| {
+            anyhow!(
+				"[verify_parachain_headers_with_grandpa_finality_proof] Invalid ancestry (base -> latest relay block)!"
+			)
+        })?;
+    }
+
+    let mut finalized = headers.ancestry(from, target.hash()).map_err(|_| {
+        anyhow!("[verify_parachain_headers_with_grandpa_finality_proof] Invalid ancestry!")
+    })?;
+    finalized.sort();
+
+    // 2. verify justification.
+    justification.verify::<Host>(client_state.current_set_id, &client_state.current_authorities)?;
+
+    // 3. verify state proofs of parachain headers in finalized relay chain headers.
```

Comment:
We'll need to refactor this to support verification of headers from different parachains